### ISSUE TITLE
Sort Mutable Buffer Chunks before moving to Read Buffer

### ISSUE THREAD

**Issue by e-dard** [influxdata/influxdb_iox#1216]
### Background

Currently chunks that are closed and moved to the read buffer are not sorted in any particular order. Instead all the tables in the MUB's chunk are just converted into record batches:

https://github.com/influxdata/influxdb_iox/blob/f290d05df02007d04da3a5473ee8a57d57e60852/mutable_buffer/src/table.rs#L354-L362

It is likely that this is resulting in poorly ordered data  in the Read Buffer, which will impact the compression and have a knock on effect on query performance in some cases. Certainly it will reduce the amount of data held we can hold in the Buffer.

### Proposal 

Throughout my design and testing of the Read Buffer I have had good success with sticking to the following approach for monitoring data workloads:

 - sort table by tag columns from least cardinality to highest cardinality.
 - if cardinality > threshold, e.g., 1,000,000 don't bother sorting.
 - do not sort field columns.
 - finally, sort the timestamp column.

I think we would benefit from applying such an approach to the MUB as the **initial default**. Obviously we will want to make the sorting of data more sophisticated and very likely operator driven (or overridden) but for now I think this is a good starting point.

### Implementation

I can see three ways to implement this, but I think one of them is simpler than the others.

1) Sort the MUB data structures before calling `to_arrow_impl` and turning into a Record Batch.
2) Convert the MUB table(s) into record batches with `to_arrow_impl` and then sort the arrays in the record batch into a new record batch using Arrow's compute kernels.
3) Get a record batch of the tables from `to_arrow_impl` for the MUB's chunk and then somehow throw them through DF to do the sorting using a `SELECT` type query.

I think (1) is tricky to do (although i probably do have `Packers` code that could do it).
I think (3) is probably complex and I don't really think it would do anything other than (2) anyway.

I think (2) is the sweet spot! Let's do that!

### But First

I don't think we should do any of this until I have determined how bad the current situation is in our internal POC. I plan to do that by getting some logging/instrumentation on the read buffer initially so I can get some detailed information about how the columns are being encoded for various tables.

- [x] https://github.com/influxdata/conductor/issues/253 determine the current situation
- [x] implement a sensible "default sort" in IOx.

**Comment by alamb** [comment]
I think #2 is a good first start. And I agree it is likely what datafusion will do today anyways

One @NGA-TRAN  and @crepererum  get basic parquet writing setup that might be an obvious source of test data for your benchmarks

**Comment by crepererum** [comment]
Sorting MIGHT also affect compression of the parquet files, depending if we wanna resort the data that comes from the ReadBuffer and is written into the parquet files.

**Comment by alamb** [comment]
@crepererum  it is a good point about sorting -- I actually think the optimal sort (and the optimal split into row groups) is likely to be quite similar for the read buffer and parquet files given their similar structures and supported encodings

**Comment by e-dard** [comment]
Good thought @crepererum. As it stand though I think having the same sort in RB and Parquet - which also lets us keep that immutable property longer, would be good. 

Optimising for the RB and then asking Parquet to just encode whatever it's given is probably a good first strategy because after all (1) parquet will probably do a good job on whatever the RB prefers; and (2) parquet data is cold and in cheap storage so we're not as sensitive to the data not being optimally compressed.

**Comment by crepererum** [comment]
Just as a side note:

> parquet data is cold and in cheap storage so we're not as sensitive to the data not being optimally compressed.

Compression (in parquet, but basically everywhere) isn't only about storage space and costs but also about transfer speeds which in many cases are longer than the actual decompression (esp. true for out-of-node storage like blob). So a well compressed parquet also means faster load times.

I totally agree with you that the sorting objective is probably very close (if not identical) for ReadBuffer and Parquet.

**Comment by NGA-TRAN** [comment]
Since Parquet data comes from RB, whatever optimized in RB will go to Parquet. I think we should keep RB and Parquet format in sync

**Comment by tustvold** [comment]
> sort table by tag columns from least cardinality to highest cardinality.

Surely the reverse of this would yield the best run length encoding?

A couple of minor notes:

* Probably want to compute the reordering, e.g. by stable sorting a list of row indexes, and then use the [take](https://docs.rs/arrow/3.0.0/arrow/compute/kernels/take/fn.take.html) kernel to reorder the actual columns
* Not necessarily part of this PR, but upsert removal should be fairly straightforward once sorted

**Comment by alamb** [comment]
@tustvold  I don't understand your question -- 

|Tag1|Tag2|time|
|--|--|--|
us-east-1|A|1000
us-east-1|B|1000
us-east-1|C|1000
us-east-1|D|1000
us-east-2|A|1001
us-east-2|B|1001

**Comment by alamb** [comment]
@tustvold  If you had the data like this:

tag1|tag2|time
|--|--|--|
us-east-1|A|1005
us-east-1|B|1002
us-east-2|C|1001
us-east-2|D|1003
us-east-2|E|1006
us-east-2|F|1004

And then sorted it like this (which I think you are suggesting)

tag1|tag2|time
|--|--|--|
us-east-2|C|1001
us-east-1|B|1002
us-east-2|F|1004
us-east-2|D|1003
us-east-1|A|1005
us-east-2|E|1006

From my reading the RLE compression for tag1 and tag2 columns would be significantly worse than the sorting in the first example

If you assume there are only distinct values of tag1 and 6 for column2 I think you can make the compression difference as large as you want. Tag1 and tag2 will take up fixed space regardless of how many rows are stored.

**Comment by tustvold** [comment]
@alamb I think we're describing the same ordering but from opposite perspectives. I was thinking at it from the perspective of the order which you perform the stable sorts, which I realise now is the opposite of how one would typically describe the resulting order :sweat_smile: So yes, we agree :grinning:

**Comment by e-dard** [comment]
I forgot to update the ticket. I observed many poorly compressing tables landing in the Read Buffer due to poorly sorted columns. It is my expectation that sorting using the default strategy I proposed above will be very beneficial.


> A couple of minor notes:
>
>  Probably want to compute the reordering, e.g. by stable sorting a list of row indexes, and then use the take kernel to > reorder the actual columns

@tustvold this sounds about right to me. I think I would:

1.  determine cardinality of tag columns, ideally from dictionary keys. Then use  that information to specify columns to  sort on.
2.  Use Arrow's `lexsort_to_index` kernel or whatever it's called  to determine the array of row IDs that specified the  correct sort.
3. Build new arrays applying sort to each input array using `take`.

👍
